### ISSUE TITLE
Import CSV file directly from the command line

### ISSUE THREAD

**Issue by mgrojo** [sqlitebrowser/sqlitebrowser#2589]
Describe the new feature
--------------------------
With a new command line option, a CSV file could be directly imported to a database, or a new in-memory database created from the data in the CSV file.

For example:
This command line would open the database my_database.db and open the import CSV dialog for the example.csv file.
```
sqlitebrowser -csv example.csv my_database.db
```

This command line would create an in-memory database and open the import CSV dialog for the example.csv file. The in-memory database could be saved as a database file afterwards.
```
sqlitebrowser -csv example.csv
```

Additional info
---------------
Please answer these questions before submitting your feature request.

#### Is your feature request related to an issue? Please include the issue number.
N/A

#### Does this feature exist in another product or project? Please provide a link.
Spreadsheet can open CSV files directly from the file explorer, because they have those files associated to the application. I'd like to have the same feature for DB4S, for the case I want to create a DB with the CSV, or simply analyze the data with the features present in DB4S and not in a spreadsheet (SQL, quick plot, etc.).

#### Do you have a screenshot? Please add screenshots to help explain your idea.
N/A

**Comment by mgrojo** [comment]
I plan to develop this, so I've written this, so it isn't forgotten, and I can receive any opinion.

**Comment by chrisjlocke** [comment]
Would it be 'automatic' (as I assume this is to import quickly with no dialog) or display the standard import dialog?  The import csv routine has a number of parameters that are different between OS's... how would you set what the delimiter was, for example? Would it just import into a 'csv' table, or would you specify that too?  What if the 'csv' table already exists - would it append or delete it?

**Comment by mgrojo** [comment]
I was thinking in opening the import dialog. LibreOffice does the same. How does Excel do? I guess, there isn't any other way, given the multiple possibilities of CSV. At least it saves some clicks.

**Comment by justinclift** [comment]
Sounds like a decent idea. :smile:

**Comment by little-brother** [comment]
> Would it be 'automatic'
The format specification can be taken through additional command line arguments. Or use the latest dialog setting for import.
P.S. Why only import? How about and export?

**Comment by mgrojo** [comment]
Ok, I think we can add two command line parameters:
`--import-csv <file.csv>` Open "Import CSV" dialog with passed file.
`--auto-import-csv <file.csv>` Automatically import passed file using last used values in the dialog.

`--import-csv` will allow associating CSV files with DB4S so they can be opened from the file manager.

`--auto-import-csv` would also allow mass import of CSV files using `--option` arguments. For example:
`--option importcsv/firstrowheader=true --option importcsv/separator=';'  --auto-import-csv file1.csv --auto-import-csv file2.csv destination.db`

**Comment by justinclift** [comment]
@mgrojo Sounds good. :smile:

@little-brother Let's start with import, and see how that goes.  No need to overdo things at the beginning. :wink:

**Comment by mgrojo** [comment]
I've implemented the first part, importing through the dialog using `--import-csv` argument.

**Comment by mgrojo** [comment]
How could I add a file association for CSV files for DB4S in Windows from our installer? I'm also trying to figure out how to do it under Linux (don't know how to provide a different command line for database files and for CSV files).

**Comment by mgrojo** [comment]
In Linux, I don't see a way to specify that we can open CSV files but using a different command line than for databases or project files, so maybe I'll need to implement detection of ASCII files, so they can be passed as normal arguments (without being preceded by `--import-csv`). I suppose SQLite and SQLCipher are always binary, aren't they?

**Comment by chrisjlocke** [comment]
> How could I add a file association for CSV files for DB4S in Windows from our installer?

I'm not convinced this is a good idea.  It's not always obvious (in Windows) what double-clicking a file will do, and you wouldn't want this to be a default option for CSV files - nine times out of ten someone would edit or view CSV files, not import.  If you double-clicked a file and it immediately started importing, it could cause problems - there would be no easy 'undo' for this (assuming it was all automatic and 'just happened').  At least via the command-line you're in control and would be in the directory you wanted the CSV file imported - it's not something that would happen 'accidentally'.
Just my thought... ;)

**Comment by mgrojo** [comment]
> > How could I add a file association for CSV files for DB4S in Windows from our installer?
> 
> I'm not convinced this is a good idea. It's not always obvious (in Windows) what double-clicking a file will do, and you wouldn't want this to be a default option for CSV files - nine times out of ten someone would edit or view CSV files, not import.

But I don't want to make it the default application for CSV files, I only want to say to the Operating System that we can open a CSV file. Then it should appear in the menu option for "Open with" or similar. See [this page](https://edu.gcfglobal.org/en/windowsbasics/opening-files-with-different-applications/1/) for Windows, or [this one](https://www.howtogeek.com/117709/how-to-change-your-default-applications-on-ubuntu-4-ways/) for Ubuntu Linux. The user will be able then to choose us to be the default application, if they want to.

> If you double-clicked a file and it immediately started importing, it could cause problems - there would be no easy 'undo' for this (assuming it was all automatic and 'just happened'). At least via the command-line you're in control and would be in the directory you wanted the CSV file imported - it's not something that would happen 'accidentally'.

Opening the CSV file using DB4S will only cause it to be imported in a new in-memory database (which you can later save as a file or not), so the CSV file is untouched, and no existent database is modified. There will be no risk on that regard.

**Comment by chrisjlocke** [comment]
>  I only want to say to the Operating System that we can open a CSV file. Then it should appear in the menu option for "Open with" or similar.

Windows 10 makes this so complex.  Windows 8 was easy - Win 10, not so.  Especially if you need to add command-line switches.  However, if the user chooses 'Open with' and 'Choose another app' then Windows will remember this, but again, sans any command-line switches.

![image](https://user-images.githubusercontent.com/7872218/108610165-6aa9cb80-73cb-11eb-84be-fef6b389463e.png)

**Comment by justinclift** [comment]
This sounds like the kind of thing @karim would know.  It's probably just some extra info we add through the Wix toolkit installer we use for Windows builds.

We've been meaning to do for other extensions anyway, so might as well add .csv in some appropriate fashion. :wink:

**Comment by mgrojo** [comment]
I'll do it without switches, otherwise I don't know how to do the file association in Linux, and probably is difficult in Windows too. I'll read, let's say, 512 bytes from the file passed as single command-line argument  and if they are only text, I'll import the file as CSV, instead of as a clear or encrypted database.

**Comment by chrisjlocke** [comment]
I'm not following the problem - why are you having to read in a portion of the file?  I thought the whole point was to read in a CSV file?

**Comment by mgrojo** [comment]
```
./sqlitebrowser my_file.my_extension
```

What are we commanding DB4S to do with my_file? :smiley: You could use file extensions, but they are not very reliable, it's better to determine in order:
* Is the file a project? Open it as project
  * Is the file a text file? Import it as a CSV file
    * Is the file a clear database? Open it
       * Try to open it as an encrypted database.

Second trial is new, the others are already implemented.

As a bonus, this will allow importing CSV files using drag&drop.

**Comment by mgrojo** [comment]
Well, I have to qualify that about drag&drop. It was already possible to drop CSV files, but only when there was a database open. Now you can drop a CSV file to import it to a new in-memory database when no one is already open.

**Comment by mgrojo** [comment]
This change is also implemented and will be available in tomorrow's build, if someone wants to test it.

Current help message:
<pre>Usage: sqlitebrowser [options] [&lt;database&gt;|&lt;project&gt;|csv-file]

Possible command line arguments:
  -h, --help          Show command line options
  -q, --quit          Exit application after running scripts
  -s, --sql &lt;file&gt;    Execute this SQL file after opening the DB
  --import-csv &lt;file&gt; Import this CSV file into the passed DB or into a new DB
  -t, --table &lt;table&gt; Browse this table after opening the DB
  -R, --read-only     Open database in read-only mode
  -S, --settings &lt;settings_file&gt;
                      Run application based on this settings file
  -o, --option &lt;group&gt;/&lt;settings&gt;=&lt;value&gt;
                      Run application with this setting temporarily set to value
  -O, --save-option &lt;group&gt;/&lt;settings&gt;=&lt;value&gt;
                      Run application saving this value for this setting
  -v, --version       Display the current version
  &lt;database&gt;          Open this SQLite database
  &lt;project&gt;           Open this project file (*.sqbpro)
  &lt;csv-file&gt;          Import this CSV file into an in-memory database
</pre>

**Comment by kos-muc** [comment]
hi mgrojo. I am new to this SQLIte world, but I do work a lot with Excel and Access and the like. I have a .ssrd database which stated that it is SQ Lite 3 at the start. SQLitebrowser opens it very well.
There is one table (category) which I need to amend with thousands of lines. (Background: it the number and the name the possible accounts (accounting framework) that I want to import into a receipt-scanning tool, i. e. base data.) What you describe here, "import csv file directly from the command line" is almost what I am looking for. The "drag and drop" feature described above would be even better. - What is the current status of this feature?

**Comment by chrisjlocke** [comment]
If you download a nightly version of DB4S (unsure of your operating system, but we have nightlies for Mac and Windows versions) then you can import a CSV file via the command-line.  Dragging/dropping isn't complete yet.
Nightlies are available from here: https://nightlies.sqlitebrowser.org/latest/

**Comment by kos-muc** [comment]
@chrisjlocke. Hi. Thx 4 your swift reply. Sorry, what is a "nightly" version? a ß-Version? - I have MS Win 10 Pro 64-bit. 
Can I install it in parallel to the current version I downloaded?
Yes, I will make sure I have a trial DB. :)
As stated, I do not have SQLite know-how. How does this work? What would be the command line?
Do I "add" to the Category table? In fact, overwriting it would be perfect (but not necessary - I can delete the few trial entries, too.)
Drag & Drop BTW sounds fantastic! :)

**Comment by chrisjlocke** [comment]
A lot to unpack here.
A nightly is an 'in development' version of DB4S - some things work, while some things don't.  Mostly things do, but users should 'tread carefully' when using it.
If you download the .zip file (32-bit or 64-bit - doesn't really matter) and extract it, you can find the .exe file buried in a host of other files.

![image](https://user-images.githubusercontent.com/7872218/109037170-55ca8200-76c2-11eb-8cba-3ebb07011165.png)

Ignore the sqlCipher version unless you need to access an encrypted database.  That version supports encryption.

WHat are you trying to accomplish?  You mention a .ssrd file with one table that you wish to 'amend' (edit or append to?)
Does the file you're trying to import contain all the columns in the category table? You don't want to delete the table if the import file only contains a small number of columns.  

You can delete all the data from a table before importing it, but just be aware that might have implications.

You say you're not familiar with SQLite, but OK with Access - I assume you're OK with relational databases in general?  Your database might have a category record which is related to another record - eg, another record points to 'category record # 15'.  If you delete all the data and import, there will no longer be a record '# 15, and instead its now # 200.

**Comment by kos-muc** [comment]
Hi Chris,

what I am trying to accomplish would be a matter of 1-2 min. in access. 

I am posting 3 screenshots. 

One is the table with the category entries (pre-set) from an expense/receipt scanning system called ScanSnap (see screenshot 2). It is very rudimentary and serves only for non-accounting, expense/receipt management purposes. 

My idea:
However, replacing (or appending, yes, this was the term :)) the "Kategorie" entity with "1234 Account name", "12345 acount name 2", "12346 account name 3" (placeholders  there are about 2000! - sample see other screenshot), will make it usable for semi-automatic accounting, as I have extracted the account framework for other purposes. :)

Yes, I did do some entity relationship modelling and was using Access - many years ago, though. I usually only work with Excel (very in-depth) and leave the "real programming" to the pros. :)

I do understand your comment regarding the record enumeration, so I will refrain from deleting it in the table. Thanks for the hint. (Clicking a couple of times more will not kill me. ;))

So, all I really need to do is to append the Category table (see screenshot) with the about 2000 entries. 
(I intend to put account number and name in the "Kategorie" column, leaving the subcategory empty, as the account framework is not hierarchical in structure and I can very easy split it again in two columns in Excel after exporting the csv file from ScanSnap once I have the required account framework in place.)

I have uploaded a small sample screenshot from the account framework (which I have generated as a structured table with much manual amending).

I hope this makes it clear?



![sample_account_framework](https://user-images.githubusercontent.com/79582870/109054861-b071d700-76de-11eb-9c07-80aabb538b8a.JPG)





![CATEGORY_table](https://user-images.githubusercontent.com/79582870/109053382-eada7480-76dc-11eb-8210-b108e90653e7.JPG)
![ScanSnap_receipt](https://user-images.githubusercontent.com/79582870/109053384-eb730b00-76dc-11eb-92ba-775a22cdd522.JPG)

**Comment by kos-muc** [comment]
> A lot to unpack here.
> A nightly is an 'in development' version of DB4S - some things work, while some things don't. Mostly things do, but users should 'tread carefully' when using it.
> If you download the .zip file (32-bit or 64-bit - doesn't really matter) and extract it, you can find the .exe file buried in a host of other files.
> 
> ![image](https://user-images.githubusercontent.com/7872218/109037170-55ca8200-76c2-11eb-8cba-3ebb07011165.png)
> 
> Ignore the sqlCipher version unless you need to access an encrypted database. That version supports encryption.
> 
> WHat are you trying to accomplish? You mention a .ssrd file with one table that you wish to 'amend' (edit or append to?)
> Does the file you're trying to import contain all the columns in the category table? You don't want to delete the table if the import file only contains a small number of columns.
> 
> You can delete all the data from a table before importing it, but just be aware that might have implications.
> 
> You say you're not familiar with SQLite, but OK with Access - I assume you're OK with relational databases in general? Your database might have a category record which is related to another record - eg, another record points to 'category record # 15'. If you delete all the data and import, there will no longer be a record '# 15, and instead its now # 200.

@chrisjlocke: I tried to answer your questions, I hope...

**Comment by kos-muc** [comment]
@mgrojo 
> Well, I have to qualify that about drag&drop. It was already possible to drop CSV files, but only when there was a database open. Now you can drop a CSV file to import it to a new in-memory database when no one is already open.>

I just read your commentary. Are you saying that it is already possible to drag & drop a csv file (that has the same columns/fields as a table in the DB) when that Table is open with the SQLitebrowser? (see second image above from 2 days ago).

**Comment by mgrojo** [comment]
> @mgrojo
> 
> > Well, I have to qualify that about drag&drop. It was already possible to drop CSV files, but only when there was a database open. Now you can drop a CSV file to import it to a new in-memory database when no one is already open.>
> 
> I just read your commentary. Are you saying that it is already possible to drag & drop a csv file (that has the same columns/fields as a table in the DB) when that Table is open with the SQLitebrowser? (see second image above from 2 days ago).

You have to drop it outside the table browser or any table or text field, otherwise it will be read as data for that field. Drop it in the toolbar, for example, and in the Import dialog, you will be able to choose the table. You can do the same using the menu: File > Import > Table from CSV file.

**Comment by kos-muc** [comment]
Thanks, @mgrojo, I have learned a lot today!  The IMPORT menu does the trick for me very well. I have managed to append the first table with the parent categories (see screenshot). The program still starts nicely. HOWEVER, the couple of marked additional items do not appear at all. - What could be wrong? Does the db have to be compiled or something before starting the program? (might be a stupid question, sorry, not a programmer)
![sqlitebrowser_2021-02-27_appended test receipt ssrd database](https://user-images.githubusercontent.com/79582870/109370384-5882ce00-78a0-11eb-8989-a27bd794b5ea.JPG)
